module/extension/Director/tests/uses_state_root.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2023 NUbots <nubots@nubots.net>
+ */
+
+#include <catch.hpp>
+#include <nuclear>
+#include <string>
+
+#include "Director.hpp"
+#include "TestBase.hpp"
+#include "util/diff_string.hpp"
+
+// Anonymous namespace to avoid name collisions
+namespace {
+
+    struct PrimaryTrigger {
+        PrimaryTrigger(const bool& run_) : run(run_) {}
+        bool run = false;
+    };
+    struct SecondaryTask {};
+    struct SubTask {};
+
+    std::vector<std::string> events;
+
+    class TestReactor : public TestBase<TestReactor> {
+    public:
+        /// Print the subtask state
+        std::string decode_run_state(GroupInfo::RunState state) {
+            switch (state) {
+                case GroupInfo::RunState::NO_TASK: return "NO_TASK";
+                case GroupInfo::RunState::RUNNING: return "RUNNING";
+                case GroupInfo::RunState::QUEUED: return "QUEUED";
+                default: return "ERROR";
+            }
+        }
+
+        explicit TestReactor(std::unique_ptr<NUClear::Environment> environment)
+            : TestBase<TestReactor>(std::move(environment)) {
+
+
+            on<Trigger<PrimaryTrigger>, Uses<SubTask>>().then(
+                [this](const PrimaryTrigger& primary, const Uses<SubTask>& subtask) {
+                    events.push_back("primary trigger subtask run state: " + decode_run_state(subtask.run_state));
+
+                    if (primary.run) {
+                        events.push_back("emitting subtask");
+                        emit<Task>(std::make_unique<SubTask>());
+                    }
+                    else {
+                        events.push_back("removing subtask");
+                        emit<Task>(std::unique_ptr<SubTask>(nullptr));
+                    }
+                });
```

Comment:
To really test the root provider stuff properly, split this into 3x reactions triggered on separate messages

One that emits the task
One that emits the nullptr
And one that just reads the uses state

That way you can check that multiple different non provider reactions all behave as if they are a single "root" provider